src/drawing/drawing_fast.cpp

```diff
@@ -0,0 +1,186 @@
+extern "C"
+{
+    #include "drawing.h"
+}
+
+// This will have -1 (0xffffffff) for (val <= 0), 0 otherwise, so it can act as a mask
+// This is expected to generate
+//     sar $0x1f, %eax (arithmetic shift right by 31)
+#define less_or_equal_zero_mask(val) (((val - 1) >> (sizeof(val) * 8 - 1)))
+
+template<int image_type, int zoom_level>
+static void FASTCALL DrawRLESprite2(const uint8* source_bits_pointer,
+                                      uint8* dest_bits_pointer,
+                                      const uint8* palette_pointer,
+                                      const rct_drawpixelinfo *dpi,
+                                      int source_y_start,
+                                      int height,
+                                      int source_x_start,
+                                      int width)
+{
+    int zoom_amount = 1 << zoom_level;
+    int zoom_mask = 0xFFFFFFFF << zoom_level;
+    uint8* next_dest_pointer = dest_bits_pointer;
+
+    int line_width = (dpi->width >> zoom_level) + dpi->pitch;
+
+    const int source_y_start_mask = less_or_equal_zero_mask(source_y_start - 1);
+    source_y_start += zoom_amount & source_y_start_mask;
+    next_dest_pointer += line_width & source_y_start_mask;
+    height -= zoom_amount & source_y_start_mask;
+
+    //For every line in the image
+    for (int i = 0; i < height; i += zoom_amount) {
+        int y = source_y_start + i;
+        uint8 i2 = i >> zoom_level;
+
+        //The first part of the source pointer is a list of offsets to different lines
+        //This will move the pointer to the correct source line.
+        const uint8 *next_source_pointer = source_bits_pointer + ((uint16*)source_bits_pointer)[y];
+
+        uint8 last_data_line = 0;
+
+        //For every data section in the line
+        while (!last_data_line) {
+            const uint8* source_pointer = next_source_pointer;
+            uint8* dest_pointer = next_dest_pointer + line_width * i2;
```

Comment:
You have added a multiplication to inside a loop... is this really an optimisation?